physical/cockroachdb.go

```diff
@@ -0,0 +1,195 @@
+		" (path STRING, value BYTES, PRIMARY KEY (path))"
+	if _, err := db.Exec(createQuery); err != nil {
+		return nil, fmt.Errorf("failed to create mysql table: %v", err)
+	}
+
+	// Setup the backend
+	c := &CockroachDBBackend{
+		table:  dbTable,
+		client: db,
+		rawStatements: map[string]string{
+			"put": "INSERT INTO " + dbTable + " VALUES($1, $2)" +
+				" ON CONFLICT (path) DO " +
+				" UPDATE SET (path, value) = ($1, $2)",
+			"get":    "SELECT value FROM " + dbTable + " WHERE path = $1",
+			"delete": "DELETE FROM " + dbTable + " WHERE path = $1",
+			"list":   "SELECT path FROM " + dbTable + " WHERE path LIKE $1",
+		},
+		statements: make(map[string]*sql.Stmt),
+		logger:     logger,
+	}
+
+	// Prepare all the statements required
+	for name, query := range c.rawStatements {
+		if err := c.prepare(name, query); err != nil {
+			return nil, err
+		}
+	}
+	return c, nil
+}
+
+// prepare is a helper to prepare a query for future execution
+func (c *CockroachDBBackend) prepare(name, query string) error {
+	stmt, err := c.client.Prepare(query)
+	if err != nil {
+		return fmt.Errorf("failed to prepare '%s': %v", name, err)
+	}
+	c.statements[name] = stmt
+	return nil
+}
+
+// Put is used to insert or update an entry.
+func (c *CockroachDBBackend) Put(entry *Entry) error {
+	defer metrics.MeasureSince([]string{"cockroachdb", "put"}, time.Now())
+
+	_, err := c.statements["put"].Exec(entry.Key, entry.Value)
+	return err
+}
+
+// Get is used to fetch and entry.
+func (c *CockroachDBBackend) Get(key string) (*Entry, error) {
+	defer metrics.MeasureSince([]string{"cockroachdb", "get"}, time.Now())
+
+	var result []byte
+	err := c.statements["get"].QueryRow(key).Scan(&result)
+	if err == sql.ErrNoRows {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	ent := &Entry{
+		Key:   key,
+		Value: result,
+	}
+	return ent, nil
+}
+
+// Delete is used to permanently delete an entry
+func (c *CockroachDBBackend) Delete(key string) error {
+	defer metrics.MeasureSince([]string{"cockroachdb", "delete"}, time.Now())
+
+	_, err := c.statements["delete"].Exec(key)
+	return err
+}
+
+// List is used to list all the keys under a given
+// prefix, up to the next prefix.
+func (c *CockroachDBBackend) List(prefix string) ([]string, error) {
+	defer metrics.MeasureSince([]string{"cockroachdb", "list"}, time.Now())
+
+	likePrefix := prefix + "%"
+	rows, err := c.statements["list"].Query(likePrefix)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var keys []string
+	for rows.Next() {
+		var key string
+		err = rows.Scan(&key)
+		if err != nil {
+			return nil, fmt.Errorf("failed to scan rows: %v", err)
+		}
+
+		key = strings.TrimPrefix(key, prefix)
+		if i := strings.Index(key, "/"); i == -1 {
+			// Add objects only from the current 'folder'
+			keys = append(keys, key)
+		} else if i != -1 {
+			// Add truncated 'folder' paths
+			keys = appendIfMissing(keys, string(key[:i+1]))
+		}
+	}
+
+	sort.Strings(keys)
```

Comment:
This might prove to be expensive if the number of keys being returned are huge. Do we really need to sort them?